### PR TITLE
[FLINK-19274][table] Implement SupportsReadingMetadata and SupportsWritingMetadata for sources and sinks

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This enables metadata columns in the planner. Both `SupportsReadingMetadata` and `SupportsWritingMetadata` are fully supported for sources and sinks.

## Brief change log

- Reimplement `CatalogTable` to `RelNode` logic for sources.
- Reimplement `CatalogTable` to `RelNode` logic for sinks.
- Support `SupportsReadingMetadata`.
- Support `SupportsWritingMetadata`.
- Support `SupportsReadingMetadata` together with `SupportsProjectionPushDown`.
- Update `TestValuesTableFactory` to support the new interfaces.
- Several exception and code improvements

## Verifying this change

This change added tests and can be verified as follows:

- `TableScanTest`
- `TableSinkTest`
- `TableSourceITCase`
- `TableSinkITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSinkTest.scala [hunk 1]
```diff
@@ -306,4 +301,101 @@ class TableSinkTest extends TableTestBase {
...
+        |SELECT TIMESTAMP '1990-10-14 06:00:00.000'
+        |""".stripMargin
+    val stmtSet = util.tableEnv.createStatementSet()
+    stmtSet.addInsertSql(sql)
+
+    thrown.expect(classOf[ValidationException])
+    thrown.expectMessage(
+      "Invalid data type for metadata key 'metadata_1' in column 'metadata_1' of table " +
+      "'default_catalog.default_database.MetadataTable'. The declared type 'TIMESTAMP(3)' cannot " +
+      "be cast to consumed metadata type 'BOOLEAN'.")
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableSinkTest.scala:366]: It looks confusing what is the "consumed" mean for users. Can we say something like `The 'metadata_1' metadata doesn't support to be declared as 'TIMESTAMP(3)', but you can declare it as types castable with 'BOOLEAN'.` ?

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableScanTest.scala [hunk 2]
```diff
@@ -76,6 +76,27 @@ class TableScanTest extends TableTestBase {
     util.verifyPlan("SELECT * FROM t1")
```
Comment (wuchong) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableScanTest.scala:76]: Could you add plan tests for projection pushdown and metadata pushdown?

**Review by wuchong — COMMENTED** [review]
Thanks for the great work @twalthr and sorry for the late response. 
The pull request looks good in general to me. I only left some minor comments.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java [hunk 3]
```diff
@@ -262,6 +266,25 @@ public DataType toPhysicalRowDataType() {
...
+	 *
+	 * <p>This method returns the query-to-sink schema.
+	 *
+	 * <p>Note: Computed columns and virtual columns are excluded in the returned row data type.
+	 *
+	 * @see DataTypes#ROW(Field...)
+	 * @see #toRowDataType()
+	 * @see #toPhysicalRowDataType()
+	 */
+	public DataType toPersistedRowDataType() {
```
[the remark this answers is not in the document]
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/api/TableSchema.java:280]: I thought we should give users a method to show the `query-to-sink` schema. It is not used for connectors and also not very useful in the planner. That's why it is in a separate commit. I can remove it again.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableScanTest.scala (hunk 2 above)
Comment (twalthr) [reply flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/TableScanTest.scala:76]: Very good point! I even found a bug.

**Review by wuchong — APPROVED** [review]
Thanks for the updating. 

LGTM.

**Pull request closed without merging**